main.go

```diff
@@ -81,6 +86,69 @@ import (
 	"github.com/networkservicemesh/cmd-nsc/internal/config"
 )
 
+func getResolverAddress() (string, error) {
+	file, err := os.Open("/etc/nsm-dns-config/resolv.conf.restore")
+	if err != nil {
+		return "", err
+	}
+
+	resolverAddr := ""
+
+	scanner := bufio.NewScanner(file)
+	scanner.Split(bufio.ScanLines)
+
+	for scanner.Scan() {
+		cfgLine := strings.Split(scanner.Text(), " ")
+		if cfgLine[0] == "nameserver" {
+			resolverAddr = cfgLine[1]
+			break
+		}
+	}
+
+	return resolverAddr, nil
+}
+
+func resolveNsmConnectURL(ctx context.Context, connectURL *url.URL) (string, error) {
+	if connectURL.Scheme == "unix" {
+		return connectURL.Host, nil
+	}
+
+	// The resolv.conf is overwritten before the connection to nsmgr is established. This turns into
+	// a chicken and egg problem. Until the connection to nsmgr is established and the nsc is
+	// receives connection context to the nse, the dns proxy would not know the IP address of the
+	// upstream dns servers, hence it cannot resolve any dns names. To fix this problem, we will read the
+	// IP address of kube-dns service backed up in /etc/nsm-dns-config/resolv.conf.restore and use it to
+	// resolve the tcp connect URL.
+	resolverAddr, err := getResolverAddress()
+	if err != nil {
+		return "", err
+	}
+
+	resolver := net.Resolver{
+		PreferGo: true,
+		Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+			dialer := net.Dialer{}
+			return dialer.DialContext(ctx, "udp", net.JoinHostPort(resolverAddr, "53"))
+		},
+	}
+
+	host, port, err := net.SplitHostPort(connectURL.Host)
+	if err != nil {
+		return "", err
+	}
+
+	addrs, err := resolver.LookupHost(ctx, host)
+	if err != nil {
+		return "", err
+	}
+
+	if len(addrs) == 0 {
+		return "", errors.New("error resolving connect URL, addr list empty")
+	}
+
+	return net.JoinHostPort(addrs[0], port), nil
+}
+
 func main() {
 	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
Do we need to have any timeout for the context

Reply:
This is just a top level ctx mainly used for initialization. A different ctx with timeout is used for server connections. Please see line 299:
dialCtx, cancelDial := context.WithTimeout(signalCtx, c.DialTimeout)